File: MinHashGenome.java
import java.io.*;
import java.util.*;
import java.nio.file.*;

// Tool to compare bacterial genomes using minhash sketches
// No alignment needed - just hashing kmers and comparing
// Usage: java MinHashGenome genome1.fasta genome2.fasta ...

public class MinHashGenome {

    static int K = 21;        // kmer length, 21 is standard for bacterial genomes
    static int S = 1000;      // sketch size

    public static void main(String[] args) throws IOException {

        if (args.length < 2) {
            System.out.println("Need at least 2 fasta files to compare");
            System.out.println("Usage: java MinHashGenome file1.fasta file2.fasta ...");
            return;
        }

        List<String> files = new ArrayList<>();
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--k")) {
                K = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--s")) {
                S = Integer.parseInt(args[++i]);
            } else {
                files.add(args[i]);
            }
        }

        System.out.println("k=" + K + " sketch_size=" + S);

        List<long[]> sketches = new ArrayList<>();
        List<String> names = new ArrayList<>();

        for (String f : files) {
            String name = f.replaceAll(".*/", "").replaceAll("\\.fna|\\.fasta", "");
            System.out.print("reading " + name + "... ");

            String seq = readFasta(f);
            long[] sk = minhash(seq);
            sketches.add(sk);
            names.add(name);
            System.out.println(seq.length() + " bp");
        }

        // print distance matrix
        System.out.print("\t");
        for (String n : names) System.out.print(n + "\t");
        System.out.println();

        for (int i = 0; i < names.size(); i++) {
            System.out.print(names.get(i) + "\t");
            for (int j = 0; j < names.size(); j++) {
                double dist = 1.0 - jaccard(sketches.get(i), sketches.get(j));
                System.out.printf("%.3f\t", dist);
            }
            System.out.println();
        }
    }

    // read fasta and return sequence as one big string
    static String readFasta(String path) throws IOException {
        StringBuilder seq = new StringBuilder();
        BufferedReader br = new BufferedReader(new FileReader(path));
        String line;
        while ((line = br.readLine()) != null) {
            line = line.trim();
            if (line.startsWith(">") || line.isEmpty()) continue;
            seq.append(line.toUpperCase());
        }
        br.close();
        return seq.toString();
    }

    // build minhash sketch from sequence
    // keep S smallest hash values seen across all kmers
    static long[] minhash(String seq) {
        // max-heap so we can efficiently track the S smallest values
        PriorityQueue<Long> heap = new PriorityQueue<>(Collections.reverseOrder());

        for (int i = 0; i <= seq.length() - K; i++) {
            String kmer = seq.substring(i, i + K);
            if (kmer.contains("N")) continue;

            // use canonical kmer (smaller of kmer and reverse complement)
            String canon = canonical(kmer);
            long h = hash(canon);

            if (heap.size() < S) {
                heap.add(h);
            } else if (h < heap.peek()) {
                heap.poll();
                heap.add(h);
            }
        }

        long[] sketch = new long[heap.size()];
        int i = 0;
        for (long h : heap) sketch[i++] = h;
        Arrays.sort(sketch);
        return sketch;
    }

    // return whichever is smaller: kmer or its reverse complement
    // this way ACGT and its rc map to the same thing
    static String canonical(String kmer) {
        String rc = revcomp(kmer);
        return kmer.compareTo(rc) < 0 ? kmer : rc;
    }

    static String revcomp(String s) {
        char[] rc = new char[s.length()];
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(s.length() - 1 - i);
            if      (c == 'A') rc[i] = 'T';
            else if (c == 'T') rc[i] = 'A';
            else if (c == 'C') rc[i] = 'G';
            else if (c == 'G') rc[i] = 'C';
            else               rc[i] = 'N';
        }
        return new String(rc);
    }

    // FNV-1a hash, works well for short strings like kmers
    static long hash(String s) {
        long h = 0xcbf29ce484222325L;
        for (int i = 0; i < s.length(); i++) {
            h ^= s.charAt(i);
            h *= 0x00000100000001B3L;
        }
        return h;
    }

    // estimate jaccard similarity by merging the two sorted sketch arrays
    // count how many of the S smallest values appear in both
    static double jaccard(long[] a, long[] b) {
        int size = Math.min(S, Math.min(a.length, b.length));
        int shared = 0, total = 0;
        int i = 0, j = 0;

        while (total < size && i < a.length && j < b.length) {
            if (a[i] == b[j])      { shared++; total++; i++; j++; }
            else if (a[i] < b[j]) { total++; i++; }
            else                   { total++; j++; }
        }
        while (total < size && i < a.length) { total++; i++; }
        while (total < size && j < b.length) { total++; j++; }

        return total == 0 ? 0 : (double) shared / total;
    }
}

File: minhash_genome.rs
use std::collections::BinaryHeap;
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

// minhash genome distance — no alignment, just kmer sketches
// usage: minhash_genome [--k 21] [--s 1000] genome1.fasta genome2.fasta ...

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();

    if args.len() < 2 {
        eprintln!("usage: minhash_genome [--k <int>] [--s <int>] file1.fasta file2.fasta ...");
        std::process::exit(1);
    }

    let mut k = 21usize;
    let mut s = 1000usize;
    let mut files = vec![];

    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--k" => { i += 1; k = args[i].parse().expect("--k needs an integer"); }
            "--s" => { i += 1; s = args[i].parse().expect("--s needs an integer"); }
            f     => files.push(f.to_string()),
        }
        i += 1;
    }

    eprintln!("k={k} s={s}");

    let (sketches, names): (Vec<_>, Vec<_>) = files.iter().map(|path| {
        let name = Path::new(path)
            .file_name().unwrap().to_str().unwrap()
            .trim_end_matches(".fasta")
            .trim_end_matches(".fna")
            .to_string();

        let seq = read_fasta(path).unwrap_or_else(|e| panic!("couldn't read {path}: {e}"));
        eprintln!("  {name}: {} bp", seq.len());

        (sketch(&seq, k, s), name)
    }).unzip();

    // distance matrix
    print!("\t");
    for n in &names { print!("{n}\t"); }
    println!();

    for i in 0..names.len() {
        print!("{}\t", names[i]);
        for j in 0..names.len() {
            print!("{:.3}\t", 1.0 - jaccard(&sketches[i], &sketches[j], s));
        }
        println!();
    }
}

fn read_fasta(path: &str) -> std::io::Result<String> {
    let mut seq = String::new();
    for line in BufReader::new(File::open(path)?).lines() {
        let line = line?;
        let line = line.trim();
        if !line.starts_with('>') && !line.is_empty() {
            seq.push_str(&line.to_uppercase());
        }
    }
    Ok(seq)
}

fn sketch(seq: &str, k: usize, s: usize) -> Vec<u64> {
    let bytes = seq.as_bytes();
    let mut heap: BinaryHeap<u64> = BinaryHeap::new(); // max-heap; tracks s smallest

    for w in bytes.windows(k) {
        if w.contains(&b'N') { continue; }

        let h = fnv1a(&canonical(w));

        if heap.len() < s {
            heap.push(h);
        } else if h < *heap.peek().unwrap() {
            heap.pop();
            heap.push(h);
        }
    }

    let mut out: Vec<u64> = heap.into_vec();
    out.sort_unstable();
    out
}

fn canonical(kmer: &[u8]) -> Vec<u8> {
    let rc = revcomp(kmer);
    if kmer <= rc.as_slice() { kmer.to_vec() } else { rc }
}

fn revcomp(seq: &[u8]) -> Vec<u8> {
    seq.iter().rev().map(|&b| match b {
        b'A' => b'T', b'T' => b'A',
        b'C' => b'G', b'G' => b'C',
        _    => b'N',
    }).collect()
}

fn fnv1a(s: &[u8]) -> u64 {
    s.iter().fold(0xcbf29ce484222325u64, |h, &b| {
        h.wrapping_mul(0x00000100000001B3) ^ b as u64
    })
}

// merge-walk the two sorted sketches and count shared hashes
fn jaccard(a: &[u64], b: &[u64], s: usize) -> f64 {
    let limit = s.min(a.len()).min(b.len());
    let (mut shared, mut total, mut i, mut j) = (0, 0, 0, 0);

    while total < limit && i < a.len() && j < b.len() {
        match a[i].cmp(&b[j]) {
            std::cmp::Ordering::Equal   => { shared += 1; total += 1; i += 1; j += 1; }
            std::cmp::Ordering::Less    => { total += 1; i += 1; }
            std::cmp::Ordering::Greater => { total += 1; j += 1; }
        }
    }
    total += (limit - total).min(a.len() - i);
    total += (limit - total).min(b.len() - j);

    if total == 0 { 0.0 } else { shared as f64 / total as f64 }
}
